File: src/core/structures/radix_trie.py
"""High-performance, memory-efficient Radix Tree (Patricia Trie) implementation.

Provides compressed-edge prefix indexing, longest prefix matching, and fast autocomplete.
Unified across CTI taxonomy lookups (MITRE ATT&CK, CWE, CVE) and search query suggestions.
"""

from __future__ import annotations

from typing import Dict, Generic, Iterator, List, Optional, Tuple, TypeVar

T = TypeVar("T")

MAX_RADIX_KEY_LENGTH = 4096
DEFAULT_SUGGEST_LIMIT = 20
MAX_SUGGEST_LIMIT = 1000


def _common_prefix_len(str1: str, str2: str) -> int:
    """Calculate length of longest common prefix between two strings."""
    limit = min(len(str1), len(str2))
    idx = 0
    while idx < limit and str1[idx] == str2[idx]:
        idx += 1
    return idx


class RadixNode(Generic[T]):
    """Internal node in Radix Tree with compressed edge string."""

    def __init__(self, prefix: str = "") -> None:
        self.prefix: str = prefix
        self.children: Dict[str, RadixNode[T]] = {}
        self.is_terminal: bool = False
        self.value: Optional[T] = None

    def add_child(self, child: RadixNode[T]) -> None:
        """Register a child node mapped by its first character."""
        if not child.prefix:
            return
        self.children[child.prefix[0]] = child

    def remove_child(self, first_char: str) -> None:
        """Remove a child by its indexing character."""
        self.children.pop(first_char, None)


def _split_edge(
    parent: RadixNode[T],
    child: RadixNode[T],
    common_len: int,
) -> RadixNode[T]:
    """Split an existing edge by inserting an intermediate node."""
    split_node: RadixNode[T] = RadixNode(child.prefix[:common_len])
    parent.add_child(split_node)

    child.prefix = child.prefix[common_len:]
    split_node.add_child(child)
    return split_node


def _insert_leaf(curr: RadixNode[T], rem_key: str, value: T) -> None:
    """Create and attach a new terminal leaf node."""
    new_leaf: RadixNode[T] = RadixNode(rem_key)
    new_leaf.is_terminal = True
    new_leaf.value = value
    curr.add_child(new_leaf)


def _advance_insert_step(
    curr: RadixNode[T], child: RadixNode[T], rem_key: str
) -> Tuple[RadixNode[T], int]:
    """Determine next node in insertion path, performing edge split if needed."""
    c_len = _common_prefix_len(rem_key, child.prefix)
    if c_len < len(child.prefix):
        split = _split_edge(curr, child, c_len)
        return split, c_len
    return child, c_len


def _step_find_node(curr: RadixNode[T], rem: str) -> Optional[Tuple[RadixNode[T], str]]:
    """Advance exact search step if prefix matches."""
    child = curr.children.get(rem[0])
    if child is None or not rem.startswith(child.prefix):
        return None
    return child, rem[len(child.prefix) :]


def _step_prefix_match(
    curr: RadixNode[T], rem: str
) -> Tuple[Optional[RadixNode[T]], Optional[str], Optional[str]]:
    """Advance prefix navigation step.

    Returns:
        (next_node, next_rem, early_match_rest)
    """
    child = curr.children.get(rem[0])
    if child is None:
        return None, None, None

    if len(rem) <= len(child.prefix):
        if child.prefix.startswith(rem):
            return child, None, child.prefix
        return None, None, None

    if not rem.startswith(child.prefix):
        return None, None, None

    return child, rem[len(child.prefix) :], child.prefix


def _push_reverse_children(
    stack: List[Tuple[RadixNode[T], str]], curr: RadixNode[T], path_acc: str
) -> None:
    """Push children to stack in reverse sorted order for deterministic forward order."""
    for char_key in sorted(curr.children.keys(), reverse=True):
        ch_node = curr.children[char_key]
        stack.append((ch_node, path_acc + ch_node.prefix))


def _record_terminal_entry(
    results: List[Tuple[str, T]], curr: RadixNode[T], path_acc: str
) -> None:
    """Append entry if node is terminal and holds a value."""
    if curr.is_terminal and curr.value is not None:
        results.append((path_acc, curr.value))


def _step_longest_match(
    curr: RadixNode[T], rem: str
) -> Optional[Tuple[RadixNode[T], str, str]]:
    """Advance longest prefix step if prefix matches."""
    child = curr.children.get(rem[0])
    if child is None or not rem.startswith(child.prefix):
        return None
    return child, rem[len(child.prefix) :], child.prefix


def _update_best_prefix_match(
    curr: RadixNode[T], accum: List[str]
) -> Optional[Tuple[str, T]]:
    """Return key-value match if current node is a valid terminal."""
    if curr.is_terminal and curr.value is not None:
        return "".join(accum), curr.value
    return None


class RadixTrie(Generic[T]):
    """Radix Tree (Compressed Trie) providing O(K) prefix operations."""

    def __init__(self) -> None:
        self.root: RadixNode[T] = RadixNode("")
        self._size: int = 0

    def __len__(self) -> int:
        return self._size

    def insert(self, key: str, value: T) -> None:
        """Insert or update a key-value pair in the Radix Tree."""
        if len(key) > MAX_RADIX_KEY_LENGTH:
            raise ValueError(
                f"Key length {len(key)} exceeds safe limit {MAX_RADIX_KEY_LENGTH}"
            )

        curr = self.root
        rem_key = key

        while rem_key:
            first_char = rem_key[0]
            if first_char not in curr.children:
                _insert_leaf(curr, rem_key, value)
                self._size += 1
                return

            curr, c_len = _advance_insert_step(curr, curr.children[first_char], rem_key)
            rem_key = rem_key[c_len:]

        if not curr.is_terminal:
            curr.is_terminal = True
            self._size += 1
        curr.value = value

    def get(self, key: str, default: Optional[T] = None) -> Optional[T]:
        """Retrieve value associated with exact key match."""
        node = self._find_exact_node(key)
        if node is not None and node.is_terminal:
            return node.value
        return default

    def contains(self, key: str) -> bool:
        """Check if exact key exists in the tree."""
        node = self._find_exact_node(key)
        return node is not None and node.is_terminal

    def __contains__(self, key: str) -> bool:
        return self.contains(key)

    def _find_exact_node(self, key: str) -> Optional[RadixNode[T]]:
        """Navigate to node exactly corresponding to key."""
        curr: Optional[RadixNode[T]] = self.root
        rem = key
        while curr is not None and rem:
            step = _step_find_node(curr, rem)
            if step is None:
                return None
            curr, rem = step
        return curr

    def _find_prefix_node(self, prefix: str) -> Optional[Tuple[RadixNode[T], str]]:
        """Locate root of subtree matching prefix and reconstructed key prefix."""
        curr: RadixNode[T] = self.root
        rem: Optional[str] = prefix
        accum: List[str] = []

        while rem is not None:
            child, next_rem, edge_str = _step_prefix_match(curr, rem)
            if child is None or edge_str is None:
                return None
            accum.append(edge_str)
            curr = child
            rem = next_rem

        return curr, "".join(accum)

    def find_by_prefix(
        self, prefix: str, limit: int = DEFAULT_SUGGEST_LIMIT
    ) -> List[Tuple[str, T]]:
        """Find all key-value entries starting with prefix up to limit."""
        eff_limit = max(1, min(limit, MAX_SUGGEST_LIMIT))
        matched = self._find_prefix_node(prefix)
        if matched is None:
            return []

        start_node, base_key = matched
        results: List[Tuple[str, T]] = []
        stack: List[Tuple[RadixNode[T], str]] = [(start_node, base_key)]

        while stack:
            if len(results) >= eff_limit:
                break
            curr, path_acc = stack.pop()
            _record_terminal_entry(results, curr, path_acc)
            _push_reverse_children(stack, curr, path_acc)

        return results

    def longest_prefix(self, text: str) -> Optional[Tuple[str, T]]:
        """Find the longest matching prefix key present in the tree."""
        curr: Optional[RadixNode[T]] = self.root
        rem = text
        accum: List[str] = []
        best_match: Optional[Tuple[str, T]] = None

        while curr is not None and rem:
            step = _step_longest_match(curr, rem)
            if step is None:
                break
            curr, rem, edge_str = step
            accum.append(edge_str)
            match = _update_best_prefix_match(curr, accum)
            if match is not None:
                best_match = match

        return best_match

    def items(self) -> Iterator[Tuple[str, T]]:
        """Iterate over all stored key-value pairs."""
        for entry in self.find_by_prefix("", limit=MAX_SUGGEST_LIMIT):
            yield entry

    def clear(self) -> None:
        """Clear all entries from the tree."""
        self.root = RadixNode("")
        self._size = 0

File: site/js/frameworks/radix-trie.js
/**
 * @fileoverview RadixTrie - Compressed prefix tree (Patricia/Radix Tree).
 *
 * A direct port of src/core/structures/radix_trie.py with O(K) insert,
 * exact-search, and prefix-search operations.
 *
 * Primary use case: 0ms incremental autocomplete for the search bars in the
 * enterprise console (#searchInput, #globalSearchInput).
 *
 * Security notes:
 *  - Keys are length-capped at MAX_KEY_LENGTH (512) to prevent DoS via
 *    pathologically deep trees.
 *  - Internal node maps use Map (not plain objects) to prevent Prototype
 *    Pollution attacks such as __proto__ injection.
 *  - No eval, no innerHTML, no external I/O.
 *
 * @package
 */

(function(global) {
  'use strict';

  /** @const {number} Maximum allowed key length. */
  var MAX_KEY_LENGTH = 512;

  /** @const {number} Default suggestion limit. */
  var DEFAULT_LIMIT = 10;

  /** @const {number} Hard upper cap for searchPrefix results. */
  var MAX_LIMIT = 1000;

  // ---------------------------------------------------------------------------
  // Internal helpers
  // ---------------------------------------------------------------------------

  /**
   * Returns the length of the longest common prefix between two strings.
   * @param {string} a
   * @param {string} b
   * @return {number}
   */
  function commonPrefixLen(a, b) {
    var len = Math.min(a.length, b.length);
    for (var i = 0; i < len; i++) {
      if (a[i] !== b[i]) return i;
    }
    return len;
  }

  // ---------------------------------------------------------------------------
  // RadixNode (internal)
  // ---------------------------------------------------------------------------

  /**
   * @constructor
   * @param {string} prefix
   */
  function RadixNode(prefix) {
    /** @type {string} */
    this.prefix = prefix;
    /** @type {boolean} */
    this.isTerminal = false;
    /** @type {*} */
    this.value = undefined;
    /** @type {!Map<string, !RadixNode>} Keyed by first character of child prefix. */
    this.children = new Map();
  }

  // ---------------------------------------------------------------------------
  // RadixTrie (public)
  // ---------------------------------------------------------------------------

  /**
   * RadixTrie is a compressed prefix tree for O(K) prefix operations.
   * All keys are strings; values are arbitrary.
   *
   * @constructor
   * @struct
   * @final
   */
  function RadixTrie() {
    /** @private {!RadixNode} */
    this.root_ = new RadixNode('');
    /** @private {number} */
    this.size_ = 0;
  }

  /**
   * Returns the number of unique keys stored.
   * @return {number}
   */
  RadixTrie.prototype.size = function() {
    return this.size_;
  };

  /**
   * Inserts or updates a key-value pair.
   * @param {string} key
   * @param {*} value
   */
  RadixTrie.prototype.insert = function(key, value) {
    if (typeof key !== 'string') {
      throw new TypeError('[RadixTrie] key must be a string.');
    }
    if (key.length > MAX_KEY_LENGTH) {
      throw new RangeError('[RadixTrie] key exceeds MAX_KEY_LENGTH (' + MAX_KEY_LENGTH + ').');
    }

    var curr = this.root_;
    var rem = key;

    while (rem.length > 0) {
      var firstChar = rem[0];
      if (!curr.children.has(firstChar)) {
        // No matching child — create a new leaf
        var leaf = new RadixNode(rem);
        leaf.isTerminal = true;
        leaf.value = value;
        curr.children.set(firstChar, leaf);
        this.size_++;
        return;
      }

      var child = curr.children.get(firstChar);
      var cLen = commonPrefixLen(rem, child.prefix);

      if (cLen < child.prefix.length) {
        // Need to split the existing edge
        var splitNode = new RadixNode(child.prefix.slice(0, cLen));
        child.prefix = child.prefix.slice(cLen);
        splitNode.children.set(child.prefix[0], child);
        curr.children.set(firstChar, splitNode);
        curr = splitNode;
      } else {
        curr = child;
      }
      rem = rem.slice(cLen);
    }

    // curr is the node for the full key
    if (!curr.isTerminal) {
      curr.isTerminal = true;
      this.size_++;
    }
    curr.value = value;
  };

  /**
   * Retrieves the value associated with an exact key, or undefined if not found.
   * @param {string} key
   * @return {*}
   */
  RadixTrie.prototype.get = function(key) {
    var node = this.findExactNode_(key);
    if (node !== null && node.isTerminal) return node.value;
    return undefined;
  };

  /**
   * Returns true if the exact key exists.
   * @param {string} key
   * @return {boolean}
   */
  RadixTrie.prototype.contains = function(key) {
    var node = this.findExactNode_(key);
    return node !== null && node.isTerminal;
  };

  /**
   * Finds all key-value entries whose keys start with prefix.
   * @param {string} prefix
   * @param {number=} limit  Maximum results (default: 10, cap: 1000).
   * @return {!Array<!Array>}  Array of [key, value] pairs.
   */
  RadixTrie.prototype.searchPrefix = function(prefix, limit) {
    var effLimit = Math.min(
        Math.max(1, typeof limit === 'number' ? limit : DEFAULT_LIMIT),
        MAX_LIMIT);

    var found = this.findPrefixNode_(prefix);
    if (found === null) return [];

    var startNode = found[0];
    var baseKey = found[1];

    var results = [];
    // Iterative DFS via explicit stack
    var stack = [[startNode, baseKey]];

    while (stack.length > 0 && results.length < effLimit) {
      var top = stack.pop();
      var curr = top[0];
      var pathAcc = top[1];

      if (curr.isTerminal && curr.value !== undefined) {
        results.push([pathAcc, curr.value]);
      }

      // Push children in reverse-sorted order for deterministic forward output
      var childKeys = Array.from(curr.children.keys()).sort().reverse();
      for (var i = 0; i < childKeys.length; i++) {
        var ch = curr.children.get(childKeys[i]);
        stack.push([ch, pathAcc + ch.prefix]);
      }
    }

    return results;
  };

  /**
   * Finds the longest prefix of text that exists as a key in the trie.
   * Returns [key, value] or null.
   * @param {string} text
   * @return {?Array}
   */
  RadixTrie.prototype.longestPrefix = function(text) {
    var curr = this.root_;
    var rem = text;
    var accum = '';
    var best = null;

    while (curr !== null && rem.length > 0) {
      var firstChar = rem[0];
      var child = curr.children.get(firstChar);
      if (!child || !rem.startsWith(child.prefix)) break;
      accum += child.prefix;
      rem = rem.slice(child.prefix.length);
      curr = child;
      if (curr.isTerminal && curr.value !== undefined) {
        best = [accum, curr.value];
      }
    }

    return best;
  };

  /**
   * Deletes a key from the trie. Returns true if the key was found and removed.
   * @param {string} key
   * @return {boolean}
   */
  RadixTrie.prototype.delete = function(key) {
    return this.deleteHelper_(this.root_, key);
  };

  /**
   * Recursive helper for delete.
   * @param {!RadixNode} curr
   * @param {string} rem
   * @return {boolean}
   * @private
   */
  RadixTrie.prototype.deleteHelper_ = function(curr, rem) {
    if (rem.length === 0) {
      if (!curr.isTerminal) return false;
      curr.isTerminal = false;
      curr.value = undefined;
      this.size_--;
      return true;
    }

    var firstChar = rem[0];
    var child = curr.children.get(firstChar);
    if (!child || !rem.startsWith(child.prefix)) return false;

    var deleted = this.deleteHelper_(child, rem.slice(child.prefix.length));
    if (!deleted) return false;

    // Prune empty non-terminal leaf nodes
    if (!child.isTerminal && child.children.size === 0) {
      curr.children.delete(firstChar);
    } else if (!child.isTerminal && child.children.size === 1) {
      // Merge single-child node
      var onlyChild = child.children.values().next().value;
      child.prefix = child.prefix + onlyChild.prefix;
      child.isTerminal = onlyChild.isTerminal;
      child.value = onlyChild.value;
      child.children = onlyChild.children;
    }
    return true;
  };

  /**
   * Clears all entries.
   */
  RadixTrie.prototype.clear = function() {
    this.root_ = new RadixNode('');
    this.size_ = 0;
  };

  /**
   * Returns all key-value pairs as an array of [key, value].
   * @return {!Array<!Array>}
   */
  RadixTrie.prototype.entries = function() {
    return this.searchPrefix('', MAX_LIMIT);
  };

  /**
   * Navigate to node exactly matching key.
   * @param {string} key
   * @return {?RadixNode}
   * @private
   */
  RadixTrie.prototype.findExactNode_ = function(key) {
    var curr = this.root_;
    var rem = key;
    while (rem.length > 0) {
      var child = curr.children.get(rem[0]);
      if (!child || !rem.startsWith(child.prefix)) return null;
      rem = rem.slice(child.prefix.length);
      curr = child;
    }
    return curr;
  };

  /**
   * Locate the subtree root and accumulated key prefix matching a given prefix.
   * @param {string} prefix
   * @return {?Array}  [RadixNode, accumulatedKey] or null.
   * @private
   */
  RadixTrie.prototype.findPrefixNode_ = function(prefix) {
    var curr = this.root_;
    var rem = prefix;
    var accum = '';

    if (prefix.length === 0) return [curr, accum];

    while (rem.length > 0) {
      var firstChar = rem[0];
      var child = curr.children.get(firstChar);
      if (!child) return null;

      if (rem.length <= child.prefix.length) {
        // rem is a prefix of child.prefix
        if (child.prefix.startsWith(rem)) {
          accum += child.prefix;
          return [child, accum];
        }
        return null;
      }

      if (!rem.startsWith(child.prefix)) return null;

      accum += child.prefix;
      rem = rem.slice(child.prefix.length);
      curr = child;
    }

    return [curr, accum];
  };

  // ---------------------------------------------------------------------------
  // Export
  // ---------------------------------------------------------------------------
  var yuzora = global['yuzora'] = global['yuzora'] || {};
  var frameworks = yuzora['frameworks'] = yuzora['frameworks'] || {};
  frameworks['RadixTrie'] = RadixTrie;

})(window);
